content/docs/reference/design-docs/editoast-errors/_index.en.md

```diff
@@ -0,0 +1,424 @@
+
+Core errors are then "lightly" wrapped: we keep the error as a generic `serde_json::Value` that we include into a `struct CoreError` that we can augment with additional information about the request. This way, the original is preserved, forwarded to the frontend, but fits our new error paradigm.
+
+`CoreError` draft:
+
+```rust
+// in editoast_core
+
+#[derive(Debug, thiserror::Error)]
+struct CoreError {
+    /// RabbitMQ "endpoint"
+    rpc: String,
+    /// Request metadata
+    metadata: HashMap<String, String>,
+    /// The original error
+    error: serde_json::Value,
+    /// Additional context the caller of the failed Core RPC can provide
+    ///
+    /// Note named "context" on purpose to avoid confusion with the `context` field of the standard OSRD error.
+    additional_information: HashMap<String, serde_json::Value>,
+}
+```
+
+Note: the `error` field is kept as a `serde_json::Value` and not parsed (even though its format is standard) as we're not supposed to perform any kind of analysis or recovery on it. If we end up parsing it in the future, that means we need a stronger mapping between Core errors and what editoast expects. The red flag will be more obvious if we end up manipulating a JSON dict instead of a proper structure.
+
+### Implementation plan
+
+We'll need a progressive migration as this implies too much change to fit in a single PR. `EditoastError` and `ViewError` will have to cohabit for some time.
+
+#### Setup
+
+1. Create the `trait views::ViewError`
+2. Implement an `axum::IntoResponse` for `ViewError` to generate a standard OSRD error response payload
+3. Add a post-processing step to the OpenAPI generation to ensure the consistency of error status codes. More details below.
+4. Create a derive macro `ViewError` that interfaces with `thiserror::Error` API and generates:
+  * `impl ViewError`
+  * `impl utoipa::ToSchema` that returns the schema of the generated OSRD error
+5. ~~Adapt the frontend error keys collection script to also look for errors in the OpenAPI routes response section.~~ We accept a temporary desync of the error keys.
+
+We decide the status code of each `ViewError` at it's definition as it's a required data to generate the response (the status code is part of the payload). Duplicating it in the `utoipa::path` annotation is a bad idea, and extracting it from the generate `utoipa::Route` is tricky. Therefore the error has to stay the source of truth about its status code.
+
+```rust
+#[utoipa::path(
+    ...,
+    responses(
+        (status = 900, body = Computation),
+        (status = 901, body = EndpointError)
+    )
+)]
+```
+
+The trick proposed here is to set each error to a non-existent status code and harmonize everything while post-processing the OpenAPI. The information will be readily available from the error schema as a constant.
```

Comment:
I'm really worry about this hack and this post-processing... but sadly, I don't think I have a good solution for this. `axum` don't really have a way to associate status code for specific error types, and even if had such a functionality, we would still have the question of how to retrofit that to OpenApi... I'd like to be in 5 years, surely someone will have figure something out 😄

Reply:
Yeah it's bad, I couldn't think of something better either... But the alternatives are much worse (incomplete response list, no utoipa, etc), so if that works... 🤷

Reply:
I SWEAR I'M STUPID, I EVEN KNEW THAT FEATURE EXISTED!

`utoipa` covers this use case, we only have to generate the implementation of `IntoResponses` for the error type...

https://docs.rs/utoipa/latest/utoipa/derive.IntoResponses.html

![gif-by-firstandmonday](https://github.com/user-attachments/assets/d9bce70a-916e-4ca5-9c75-24cbef2e1573)

Reply:
Looks promising, we'll see during implementation. I'm resolving.